### ISSUE TITLE
Austrian License Plates

### ISSUE THREAD

**Issue by hetzenmat** [openalpr/openalpr#292]
The OpenALPR library does not work well for Austrian license plates. It classifies German plates fine but they use a different font.

The google images show the differences:
[German license plates](https://www.google.at/search?q=deutschland+kennzeichen&biw=1280&bih=913&source=lnms&tbm=isch&sa=X&sqi=2&ved=0ahUKEwiq_ZLPk6TLAhVqMJoKHfP-ByYQ_AUIBigB) vs. [Austrian license plates](https://www.google.at/search?q=austrian+license+plates&source=lnms&tbm=isch&sa=X&ved=0ahUKEwiEptmzk6TLAhVqCpoKHVvYBB8Q_AUIBygB&biw=1280&bih=913#tbm=isch&q=%C3%B6sterreich+kennzeichen)

I am running Ubuntu 14.04, installed the OpenALPR binaries and tried to train the OCR using the [train-ocr](https://github.com/openalpr/train-ocr) repository but the util program openalpr-utils-binarizefontsheet is missing. Furthermore, the program `classifychars` does not work as documented.
Is it possible to make OpenALPR support Austrian license plates?

**Comment by matthill** [comment]
Yes, that font does look different than any of the EU fonts we use for training.  Often countries have strong similarities (such that training for that particular country is not necessary) but that does not seem to be the case for Austria.  

Are you able to collect 100-200 images and crop the characters out using the utility described here:
http://doc.openalpr.com/accuracy_improvements.html#training-ocr

Alternatively, do you have a TTF font that precisely represents the fonts used on the plates?  If so, I can add the data to the training set and retrain -- that would definitely improve accuracy for Austria.

**Comment by hetzenmat** [comment]
Thanks for the quick response!

I already tried to use the openalpr-utils-binarizefontsheet to train the ocr, but the default installation of tesseract does not include all the required utilitiy programs.

On page 21 of [this PDF](https://www.ris.bka.gv.at/Dokumente/BgblPdf/2002_376_2/2002_376_2.pdf) are the official characters of Austrian license plates.

I would really appreciate if you take time and retrain for Austrian license plates so you can support one additional country!

**Comment by matthill** [comment]
Thanks for the data.  Yes, I think a font sheet could work there.  My approach would be: take a screenshot of the PDF and replicate it for more rows/columns (maybe 3 columns, 9 rows) then print it, take a few pictures, and extract the characters via the binarize font sheet utility.

**Comment by matthill** [comment]
[2002_376_2.pdf](https://github.com/openalpr/openalpr/files/157958/2002_376_2.pdf)

Uploading the PDF directly, in case the link ever goes away

**Comment by hetzenmat** [comment]
So are you able to retrain with the PDF? As I said, unfortunately, I couldn't get the train.py script of the train-ocr repository working.

**Comment by matthill** [comment]
Yes, I can add it to my 'to do' list.  It may be a while before I get to it, though.

**Comment by hetzenmat** [comment]
Thank you!

Whether I manage to train myself or you finish it first, I will close the issue.

**Comment by ghlang** [comment]
Hi, was there any progress with the austrian plates? Attached are a tif and box file of actual austrian license plate characters. Could you be so kind and run them through train.py, as it gives me some troubles too. 
[leu.austria.exp0.tif.zip](https://github.com/openalpr/openalpr/files/368043/leu.austria.exp0.tif.zip)
[leu.austria.exp0.box.zip](https://github.com/openalpr/openalpr/files/368036/leu.austria.exp0.box.zip)

**Comment by ujikol** [comment]
Hi,
Could somebody progress with Austrian plates?
Does this training also impact the recognition of plates as such or only the OCR? In my first tests alpr did not recognize a single plate. So should I expect that this can be improved by this training, or is the detection algorithm too sensitive to different view perspectives?
thx

**Comment by chrisiaut** [comment]
@matthill are we still on your 'to do' list? :)

**Comment by PetarMG-BG** [comment]
Hi guys,
I have a progress about training Austrian plates, but the files that @ghlang provided are with not good quality. Did some one have a traineddata already, or just tif files with better resolution?
I can handle the training, but I'm still not aware how to produce the tif files and also I don't have a big plate pictures pack.

**Comment by alucard079** [comment]
@PetarMG-BG I can help you in producing tif and box file I've done that correctly but can you help me too on how to produced the `traineddata` I have installed the tesseract-ocr which I'm facing right now, because when I run the `train.py` it gives me this error.

Two-Letter Country Code to Train: gb
Processing: ./gb/input/luk.uk.exp0.box
Executing: /home/nigel/tesseract-ocr/tesseract -l eng ./gb/input/luk.uk.exp0.tif luk.uk.exp0 nobatch box.train.stderr
sh: 1: /home/nigel/tesseract-ocr/tesseract: not found
mv: cannot stat './luk.uk.exp0.tr': No such file or directory
mv: cannot stat './luk.uk.exp0.txt': No such file or directory
sh: 1: /home/nigel/tesseract-ocr/training/unicharset_extractor: not found
Executing: /home/nigel/tesseract-ocr/training/mftraining -F ./tmp/font_properties -U unicharset -O ./tmp/lgb.unicharset ./tmp/*.tr
sh: 1: /home/nigel/tesseract-ocr/training/mftraining: not found
rm: cannot remove './unicharset': No such file or directory
mv: cannot stat './tmp/lgb.unicharset': No such file or directory
cp: cannot stat './gb/input/unicharambigs': No such file or directory
sh: 1: /home/nigel/tesseract-ocr/training/cntraining: not found
mv: cannot stat './shapetable': No such file or directory
mv: cannot stat './pffmtable': No such file or directory
mv: cannot stat './inttemp': No such file or directory
mv: cannot stat './normproto': No such file or directory
sh: 1: /home/nigel/tesseract-ocr/training/combine_tessdata: not found
mv: cannot stat './lgb.unicharset': No such file or directory
mv: cannot stat './lgb.shapetable': No such file or directory
mv: cannot stat './lgb.pffmtable': No such file or directory
mv: cannot stat './lgb.inttemp': No such file or directory
mv: cannot stat './lgb.normproto': No such file or directory
mv: cannot stat './lgb.unicharambigs': No such file or directory

This is my current directory. `/home/nigel/train/train-ocr`
My tesseract-ocr installed on `/home/nigel/tesseract-ocr`

this is my code in `train.py` `TESSERACT_DIR='/home/nigel/tesseract-ocr/src'`

Please help me with this.

**Comment by PetarMG-BG** [comment]
I'm using another tool for the generating the traineddata. You can find it here:
http://vietocr.sourceforge.net/training.html

**Comment by alucard079** [comment]
But can I use my generated tif and box files for this tool? Or should I start again from the scratch?

**Comment by PetarMG-BG** [comment]
@alucard079  this tool is working with tesseract, and you can use your already produced files. It's very easy to work with.

**Comment by alucard079** [comment]
@PetarMG-BG is it normal that my `z` read as `2`? Can you give me steps on how to that after importing the `tif file'. I training the philippines license plate, I don't think if the OCR language is matter when you training it. 

![51737800_536562236834945_8187146202429521920_n](https://user-images.githubusercontent.com/28977709/52399080-14c09e80-2af6-11e9-858e-ac7573a044ed.png)

**Comment by PetarMG-BG** [comment]
@alucard079 you are looking in the VietOCR, for training you have to open the jTessBoxEditor. Look at tha link that U provide you. There is a complete tutorial.

**Comment by losabio1617** [comment]
@PetarMG-BG i also doing new country data ,do i need to train new language from jTessBoxEditor for training philipine plate, i tried to used eng,leu,and eu but the ocr is not accurate

**Comment by PetarMG-BG** [comment]
@losabio1617  - yes you have to train.

**Comment by losabio1617** [comment]
@PetarMG-BG  theres a diff font format that i am training in the ph plate white plate and green plate 0 and 1 of this two plate completely diff each other ,is it possible if i will combine the language that generated from jTessBox or i will make a seperate data for each one

**Comment by losabio1617** [comment]
![screenshot from 2019-02-08 10-26-52](https://user-images.githubusercontent.com/38846946/52455654-1b015a00-2b8c-11e9-82fe-d6ce557f68bf.png)
@PetarMG-BG  i just have this error (attached image) i skipped on the Box editor and tiff/box generator because i already have box and tif file from that i generated from openalpr , i copy pasted the pathed of this two --
/home/losabio/Downloads/jTessBoxEditor-2.2.0/jTessBoxEditor/tesseract-ocr/combine_tessdata.exe
___________________________________
losabio@losabio-System-Product-Name:~/Downloads/jTessBoxEditor-2.2.0/jTessBoxEditor/tesseract-ocr/tessdata/input$ ls
lph.philippine.exp0.box  ph.font_properties      ph.words_list
lph.philippine.exp0.tif  ph.frequent_words_list  training.log
Yet it i got an error no such file or directory

**Comment by alucard079** [comment]
@PetarMG-BG what if I have two tiff files, can I combine them into one `traineddata`?

**Comment by losabio1617** [comment]
@PetarMG-BG 
A great thank you!!! im done generating the language data but what if i have two diff format should i create a box and tif with there characters or separate tif and box and combine both of them??

**Comment by PetarMG-BG** [comment]
@alucard079 and @losabio1617 - you can combine the two tiffs in one file with some image redactor, and use one of the box files. They have to be the same name.

**Comment by alucard079** [comment]
Thanks we can generate now traineddata. However we would like to combine two tif file into one traineddata. 

Can I know what is image redactor, do you have some link to that?

Thanks again!

**Comment by alucard079** [comment]
Anw, do you solved now on how to generate the box and tif file using openalpr utilities. Losabio can help you out there

**Comment by PetarMG-BG** [comment]
You can do it with whatever generator you want. I personally now coding another ALPR software. This one is not so accurate and need's a lot of data and hard training process :)

**Comment by losabio1617** [comment]
yeah i just use paint also,, is there a way to combine two diff language trained data i already generate one for the white plate in ph and the accurancy is good i want to merge it once i finish the green plate data,, we tried some application to merge  diff data but the ocr have some output of small letters, i rechecked the tif file that i used and all letters are capitals do u know whats is the reason OCR generate some small letter output?

**Comment by alucard079** [comment]
@PetarMG-BG Sir, do you tried running the alpr with ip camera? to make it realtime maybe using the "alpr -j" command or python integration using opencv?

**Comment by gkieninger** [comment]
Hallo!
I am also interested to improve OpenALPR for Austrian Plates. If I can help please tell me and on the other Hand I would be happy to get the result.
Trank you

**Comment by ooxtcoo** [comment]
do someone have made austrian plates training files and is willing to share?

if not, i can provide tausands of austrian plates with 8mp images.

**Comment by michaelradhuber** [comment]
> do someone have made austrian plates training files and is willing to share?
> 
> if not, i can provide tausands of austrian plates with 8mp images.
> so we could produce best possible result.

Send me the files I will post the training file here then

**Comment by michaelradhuber** [comment]
@aperantogalasio on Twitter, thanks

**Comment by jpunz** [comment]
@michaelradhuber I tried to generate the traineddata for Austrian license plates - unfortunately something went wrong, as alpr doesn't recognize any Austrian license plates.

Can you provide me your traineddata file?
Do config- and region-file differ from the EU?
